src/Bicep.Core/Parser/Parser.cs

```diff
@@ -29,21 +30,31 @@ public Parser(string text)
 
         public ProgramSyntax Program()
         {
-            var statements = new List<SyntaxBase>();
+            var declarations = new List<SyntaxBase>();
+            
             while (!this.IsAtEnd())
             {
                 var statement = Declaration();
-                statements.Add(statement);
+                declarations.Add(statement);
+
+                if (statement is IDeclarationSyntax)
+                {
+                    // declarations must be followed by a newline or the file must end
+                    var newLine = this.WithRecoveryNullable(this.NewLineOrEof, consumeTerminator: true, TokenType.NewLine);
+                    if (newLine != null)
+                    {
+                        declarations.Add(newLine);
+                    }
+                }
```

Comment:
What's the reason for the conditional `statement is IDeclarationSyntax`?

Reply:
Let me add a comment for this.

Reply:
Added